File: bfs & dfs/maze.py
#백준 2178 미로 탐색

from collections import deque
from sys import stdin

n, m = map(int, stdin.readline().split())

graph = []
for i in range(n):
    graph.append(list(map(int, input())))

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def bfs(x, y):
    queue = deque()
    queue.append((x, y))

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            #이동시 범위 벗어남
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            #이동시 이동할 수 없는 칸
            if graph[nx][ny] == 0:
                continue
            if graph[nx][ny] == 1:
                #거리 계산
                graph[nx][ny] = graph[x][y] + 1
                queue.append((nx, ny))
    return graph[n-1][m-1]

print(bfs(0,0))

File: bfs & dfs/maze.js
// 백준 2178번 미로 탐색

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().trim().split("\n");

const nm = input.shift().split(" ").map(Number);
const n = nm[0];
const m = nm[1];
const maze = [];

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

while (input.length !== 0) {
  let temp = input.shift().split("");
  if (temp.includes("\r")) temp.pop();
  maze.push(temp.map(Number));
}

const bfs = (x, y) => {
  let queue = [];
  queue.push([x, y]);
  while (queue.length !== 0) {
    let [x, y] = queue.shift();
    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i];
      let ny = y + dy[i];
      if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
      if (maze[nx][ny] === 0) continue;
      if (maze[nx][ny] === 1) {
        maze[nx][ny] = maze[x][y] + 1;
        queue.push([nx, ny]);
      }
    }
  }
  return maze[n - 1][m - 1];
};

console.log(bfs(0, 0));
